dNR: make sure every rule ID is unique

elijahsawyers opened the pull request:
#### 202477fe85b4db3ca4477237f6321ee444325fc4
<pre>
dNR: make sure every rule ID is unique
<a href="https://bugs.webkit.org/show_bug.cgi?id=298070">https://bugs.webkit.org/show_bug.cgi?id=298070</a>
<a href="https://rdar.apple.com/72159785">rdar://72159785</a>

Reviewed by Brian Weinstein.

This simple patch reports an error when a static dNR ruleset contains a duplicate identifier. It&apos;s
important that we do this so that developers are aware that they might be getting incorrect
information from onRuleMatchedDebug.

The error is surfaced in Safari&apos;s extensions preferences pane.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
(WebKit::for):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, DuplicatedRuleIDs)):

Canonical link: <a href="https://commits.webkit.org/299456@main">https://commits.webkit.org/299456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f471cc985317f8c1f37e95f4e496a00e98e604e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119098 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38779 "Built successfully") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/71161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/227face1-83f6-4fc6-811a-4d205ced211a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120976 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/39475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/47361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/125314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/71161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fd331d38-c863-4a9e-b521-5b2828b92683) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122051 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/39475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6a3f446d-b066-4399-a02e-f1cb5c6fed04) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/39475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/68965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/39475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/46005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/47361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/128339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/46372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/29430 "Failed to checkout and rebase branch from PR 50070") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18956 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/45875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/51555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/45342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/48688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/47027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4671,6 +4671,9 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
         auto *allConvertedRules = [_WKWebExtensionDeclarativeNetRequestTranslator translateRules:allJSONObjects errorStrings:&parsingErrorStrings];
 
+        for (NSString *errorString in parsingErrorStrings)
+            recordError(m_extension->createError(WebExtension::Error::InvalidDeclarativeNetRequest, errorString));
```
It's possible to have both errors and a valid set of rules? So if there's a duplicate rule we will just drop it on the floor?

elijahsawyers replied:
We can keep rules with duplicate IDs. There's not much of a reason to throw it away if the rest of the rule is valid.

b-weinstein replied:
Looking at the code, it appears we even keep the rule around, so never mind this comment.

elijahsawyers commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -589,6 +589,9 @@
 "Exceeded maximum number of enabled `declarative_net_request` static rulesets. The first %lu will be applied, the remaining will be ignored." = "Exceeded maximum number of enabled `declarative_net_request` static rulesets. The first %lu will be applied, the remaining will be ignored.";
 
+/* WKWebExtensionErrorInvalidDeclarativeNetRequestEntry description for a duplicated rule id */
+"`declarative_net_request` ruleset with id `%@` duplicates the rule id `%@`." = "`declarative_net_request` ruleset with id `%@` duplicates the rule id `%@`.";
```
I should actually make sure that any of the other dNR rule parsing errors returned in `_WKWebExtensionDeclarativeNetRequestRule` are localized now that we're actually reporting them.

elijahsawyers replied:
Actually, this is quite a bit of work to do. I've filed a FIXME: rdar://159673819

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -1448,6 +1448,45 @@
+    NSArray<NSError *> *errors = manager.get().context.errors;
+    ASSERT_EQ(errors.count, 1ul);
+    ASSERT_TRUE([errors.firstObject.localizedDescription isEqualToString:@"`declarative_net_request` ruleset with id `duplicated_rule_id` duplicates the rule id `1`."]);
+}
```
Is it worth writing a test that the same rule ID across different rulesets doesn't give an error?

elijahsawyers committed:
dNR: make sure every rule ID is unique
https://bugs.webkit.org/show_bug.cgi?id=298070
rdar://72159785

Reviewed by Brian Weinstein.

This simple patch reports an error when a static dNR ruleset contains a duplicate identifier. It's
important that we do this so that developers are aware that they might be getting incorrect
information from onRuleMatchedDebug.

The error is surfaced in Safari's extensions preferences pane.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm:
(WebKit::for):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIDeclarativeNetRequest, DuplicatedRuleIDs)):

Canonical link: https://commits.webkit.org/299456@main
